beacon_node/lighthouse_network/src/behaviour/gossip_cache.rs

```diff
@@ -0,0 +1,241 @@
+    pub fn voluntary_exit_timeout(mut self, timeout: Duration) -> Self {
+        self.voluntary_exit = Some(timeout);
+        self
+    }
+
+    /// Timeout for proposer slashings.
+    pub fn proposer_slashing_timeout(mut self, timeout: Duration) -> Self {
+        self.proposer_slashing = Some(timeout);
+        self
+    }
+
+    /// Timeout for attester slashings.
+    pub fn attester_slashing(mut self, timeout: Duration) -> Self {
+        self.attester_slashing = Some(timeout);
+        self
+    }
+
+    /// Timeout for aggregated sync commitee signatures.
+    pub fn signed_contribution_and_proof(mut self, timeout: Duration) -> Self {
+        self.signed_contribution_and_proof = Some(timeout);
+        self
+    }
+
+    /// Timeout for sync commitee messages.
+    pub fn sync_committee_message(mut self, timeout: Duration) -> Self {
+        self.sync_committee_message = Some(timeout);
+        self
+    }
+
+    pub fn build(self) -> GossipCache {
+        let GossipCacheBuilder {
+            default_timeout,
+            beacon_block,
+            aggregates,
+            attestation,
+            voluntary_exit,
+            proposer_slashing,
+            attester_slashing,
+            signed_contribution_and_proof,
+            sync_committee_message,
+        } = self;
+        GossipCache {
+            expirations: DelayQueue::default(),
+            topic_msgs: HashMap::default(),
+            beacon_block: beacon_block.or(default_timeout),
+            aggregates: aggregates.or(default_timeout),
+            attestation: attestation.or(default_timeout),
+            voluntary_exit: voluntary_exit.or(default_timeout),
+            proposer_slashing: proposer_slashing.or(default_timeout),
+            attester_slashing: attester_slashing.or(default_timeout),
+            signed_contribution_and_proof: signed_contribution_and_proof.or(default_timeout),
+            sync_committee_message: sync_committee_message.or(default_timeout),
+        }
+    }
+}
+
+impl GossipCache {
+    pub fn builder() -> GossipCacheBuilder {
+        GossipCacheBuilder::default()
+    }
+
+    // Insert a message to be sent later.
+    pub fn insert(&mut self, topic: GossipTopic, data: Vec<u8>) {
+        let expire_timeout = match topic.kind() {
+            GossipKind::BeaconBlock => self.beacon_block,
+            GossipKind::BeaconAggregateAndProof => self.aggregates,
+            GossipKind::Attestation(_) => self.attestation,
+            GossipKind::VoluntaryExit => self.voluntary_exit,
+            GossipKind::ProposerSlashing => self.proposer_slashing,
+            GossipKind::AttesterSlashing => self.attester_slashing,
+            GossipKind::SignedContributionAndProof => self.signed_contribution_and_proof,
+            GossipKind::SyncCommitteeMessage(_) => self.sync_committee_message,
+        };
+        let expire_timeout = match expire_timeout {
+            Some(expire_timeout) => expire_timeout,
+            None => return,
```

Comment:
Its interesting behaviour that if someone doesn't set a default and a specific timeout we drop the messages silently. 

I assume this is expected, because it allows us to customize this to ignore some kinds of messages. Just pointing it out as future devs could accidentally drop messages if they don't specify a timeout and remove the default.

Reply:
yeah it's intended. I'll add it to the docs

Reply:
docs here https://github.com/sigp/lighthouse/pull/2964/commits/80b0b229af603ea79d95ba08297366e3a8a42121